pkg/controller/pxc/controller.go

```diff
@@ -1095,3 +1058,77 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
 
 	return nil
 }
+
+func (r *ReconcilePerconaXtraDBCluster) createOrUpdate(obj runtime.Object) error {
+	metaAccessor, ok := obj.(metav1.ObjectMetaAccessor)
+	if !ok {
+		return errors.New("Can't convert object to ObjectMetaAccessor")
+	}
+
+	objectMeta := metaAccessor.GetObjectMeta()
+
+	if objectMeta.GetAnnotations() == nil {
+		objectMeta.SetAnnotations(make(map[string]string))
+	}
+
+	delete(objectMeta.GetAnnotations(), "percona.com/last_config_hash")
+
+	hash, err := getObjectHash(obj)
+	if err != nil {
+		return errors.Wrap(err, "calculate object hash")
+	}
+
+	oldObject := obj.DeepCopyObject()
+
+	err = r.client.Get(context.Background(), types.NamespacedName{
+		Name:      objectMeta.GetName(),
+		Namespace: objectMeta.GetNamespace(),
+	}, oldObject)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "get object")
+	}
+
+	objectMeta.GetAnnotations()["percona.com/last_config_hash"] = hash
+
+	if k8serrors.IsNotFound(err) {
+		return r.client.Create(context.TODO(), obj)
+	}
+
+	oldObjectMeta := oldObject.(metav1.ObjectMetaAccessor).GetObjectMeta()
+
+	if oldObjectMeta.GetAnnotations()["percona.com/last_config_hash"] != hash ||
+		!isObjectMetaEqual(objectMeta, oldObjectMeta) {
+
+		objectMeta.SetResourceVersion(oldObjectMeta.GetResourceVersion())
+		switch object := obj.(type) {
+		case *corev1.Service:
+			object.Spec.ClusterIP = oldObject.(*corev1.Service).Spec.ClusterIP
+		}
+		return r.client.Update(context.TODO(), obj)
+	}
+	return nil
+}
+
+func getObjectHash(obj runtime.Object) (string, error) {
+	var dataToMarshall interface{}
+	switch object := obj.(type) {
+	case *appsv1.StatefulSet:
+		dataToMarshall = object.Spec
+	case *appsv1.Deployment:
+		dataToMarshall = object.Spec
+	case *corev1.Service:
+		dataToMarshall = object.Spec
+	default:
+		dataToMarshall = obj
+	}
+	data, err := json.Marshal(dataToMarshall)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(data), nil
+}
+
+func isObjectMetaEqual(old, new metav1.Object) bool {
+	return reflect.DeepEqual(old.GetAnnotations(), new.GetAnnotations()) &&
+		reflect.DeepEqual(old.GetLabels(), new.GetLabels())
+}
```

Comment:
If we compare maps why do we need reflection?

Reply:
because this function doing exactly the same with maps.
In that case we don't need to write this code. Plus reflection is actively used is this pr already

Reply:
No, it does not. It try to understand what a type is it, allocate another map for tracking of visited entries, do a lot of unneeded stuff to know what we already know.

Reply:
054a623